pkg/estimator/hll/compressed.go

```diff
@@ -0,0 +1,155 @@
+package hll
+
+import "encoding/binary"
+
+// Original author of this file is github.com/clarkduvall/hyperloglog
+type iterable interface {
+	decode(i int, last uint32) (uint32, int)
+	Len() int
+	Iter() *iterator
+}
+
+type iterator struct {
+	i    int
+	last uint32
+	v    iterable
+}
+
+func (iter *iterator) Next() uint32 {
+	n, i := iter.v.decode(iter.i, iter.last)
+	iter.last = n
+	iter.i = i
+	return n
+}
+
+func (iter *iterator) Peek() uint32 {
+	n, _ := iter.v.decode(iter.i, iter.last)
+	return n
+}
+
+func (iter iterator) HasNext() bool {
+	return iter.i < iter.v.Len()
+}
+
+type compressedList struct {
+	count uint32
+	last  uint32
+	b     variableLengthList
+}
+
+func (v *compressedList) MarshalBinary() (data []byte, err error) {
+	// At least 4 bytes for the two fixed sized values.
+	data = make([]byte, 0, 4+4)
+
+	// Marshal the count and last values.
+	data = append(data, []byte{
+		// Number of items in the list.
+		byte(v.count >> 24),
+		byte(v.count >> 16),
+		byte(v.count >> 8),
+		byte(v.count),
+		// The last item in the list.
+		byte(v.last >> 24),
+		byte(v.last >> 16),
+		byte(v.last >> 8),
+		byte(v.last),
+	}...)
+
+	// Marshal the variableLengthList
+	bdata, err := v.b.MarshalBinary()
```

Comment:
You could do this before L42 and calculate the size of `data` so that append below does not need to grow the slice.